### PR TITLE
Add `CalculatorPack` (grouping of calculators/calibrators into "packs")

### LINKED ISSUE #184: Saving groups of calculators
The calculators reorganization of ixdat (#182) seeks to avoid calculator redundancy - this saw the end of `ECMSCalculator` in favor of separate `MSCalculator` and `ECCalculator`s. A loss is that it was quite convenient to save an `ECMSCalculator` and use later, rather than have to separately transfer the two types of information (sensitivity factors in one, and `RE_vs_RHE`, `A_el`, and `R_Ohm` in another). A solution (see https://github.com/ixdat/ixdat/pull/182#discussion_r1750200444) could be to save multiple calculators at once, e.g. a measurements full `calculators` dictionary, in a single statement.

### PR DESCRIPTION
**Pull request by FrederikLizakJohansen**
This PR introduces a new CalculatorPack feature, trying to address Issue #184 and following up on the original comment by @AnnaWiniwarter in #182.

### Why it would be useful

From what I gather from the discussions, in some workflows, a measurement is analyzed using several calculators at once (e.g. an ECCalibration together with an MSCalibration). Currently, if we want to apply the same calibrations to another dataset, each calculator has to be recreated or copied separately.

`CalculatorPack` lets users capture all relevant calculators from a measurement in one step and re-apply them elsewhere, making it more **convenient** to reproduce results, while also making it more **portable** (packs are plain JSON files that can be versioned, shared, or stored with raw data), and possibly **extendable** in the future.

### Overview
`CalculatorPack` is a container for a set of calculators (EC, MS, SIQ, ...) extracted from a `Measurement`. It allows them to be:

- captured from a measurement,
- serialized to JSON or database, 
- saved/loaded from files (.ixpack.json), and
- re-attached to another measurement, with optional conflict resolution.

This makes it possible to “carry over” calibration and quantification workflows between datasets.

### Changes:

1. ixdat/calculators/packs.py _(new file)_

    - Defines `CalculatorPack` with helpers for serialization, integrity checking, file/DB I/O, and re-attachment. 
    - Skips Indexer calculators (they are always created by measurements).
    - Uses the Calculators native `as_dict()`/`from_dict()` if available. Else uses lightweight serialization.

2. ixdat/measurement_base.py _(updated)_

    - Adds integration helpers to create a pack from a measurement (`save_calculators_pack`) and apply one back (`apply_calculators_pack`).

3. development_scripts/demo_calculator_pack.py _(new demo)_

    - Demonstration of capturing EC calibrations, saving/loading packs, and re-attaching them to new measurements. **note: due to missing test data on my side, I have not been able to test multiple calculator types**.

### Still Missing / To Discuss

1. Plugin handling:

    - Some calculators (e.g. `siqCalculator`) depend on external packages (`spectro_inlets_quantification`). Currently `CalculatorPack` has no way to handle this robustly.

    - An approach might be a central plugin registry where calculator classes (or their `calculator_type`) can declare the external module and optional activation function.

    - Question: would you prefer this registry approach, or to extend each calculator with something like a `requires_plugin` attribute?

2. Broader testing:

    - Unit tests are still missing for edge cases (conflict resolution, _missing plugins_, integrity failures).
    - Only a demo script exists so far.
    
    
Closes #184

### CODE REVIEW

File: development_scripts/demo_calculator_pack.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    print("\n[4] Computing calibrated EC series on src:")
+    compute_ec_series(src)
+
+    # 4) Build a CalculatorPack and write to file (exclude Indexer by default)
+    print("\n[5] Building CalculatorPack (excluding Indexer) and writing to file...")
+    pack = CalculatorPack.from_measurement(
+        src,
+        name="demo_pack",
+        notes="EC-only demo",
+        exclude=["Indexer", "ixdat.calculators.indexer.Indexer"],
```
Comment (ScottSoren) [inline development_scripts/demo_calculator_pack.py]: I tried commenting out this line to no effect. Indexer still doesn't go into the pack.
Also, why should one have to exclude both "Indexer" and "ixdat.calculators.indexer.Indexer"? - shouldn't one be enough?

File: src/ixdat/calculators/packs.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    """
+    A container that serializes multiple calculators together so they can be
+    saved/loaded and re-applied to another Measurement in a single operation.
+
+    Overview:
+    - Captures a set of calculators (MS, EC, backgrounds, siq, ...)
+      with minimal metadata and per-calculator payloads.
+    - Supports:
+        `CalculatorPack.from_measurement(...)` to capture calculators
+        `pack.to_file(path)` and `CalculatorPack.read(path=...)` for file I/O
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py]: For consistency with other ixdat methods, `to_file` should probably be either `save` (if it uses a backend) or `export` (if not)

File: src/ixdat/calculators/packs.py [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+
+        # Internal storage:
+        self._calculators: List[Any] = list(calculators) if calculators else []
+        self._items: Optional[List[JsonDict]] = None  # filled by from_json or to_json
+
+        # Backing JSON for Saveable (db column). Keep it None until first to_json()
+        self.json: Optional[str] = None
+
+    @staticmethod
+    def _serialize_calculator(cal: Any) -> dict:
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py]: Hm... could you include a sentence in the docstring about when this method is needed instead of directly using `as_dict()`?
I'm up for revisiting the whole `as_dict` serialization as part of getting ixdat a real SQL backend (e.g. the infamous #75). With that in the back of our minds, is it safe to say that having two serializations here is a temporary thing?

File: src/ixdat/calculators/packs.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+        safe: dict[str, Any] = {"__format__": "light.dict"}
+        for k, v in vars(cal).items():
+            if k == "measurement":
+                continue  # don't persist the measurement object
+            if isinstance(v, (int, float, str, list, dict, type(None))):
+                safe[k] = v
+        return safe
+
+    @staticmethod
+    def _deserialize_calculator(cls: type, payload: dict) -> Any:
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py]: Same question wrt why not just use `from_dict` directly, or in the long run merge these two deserialization methods.

File: src/ixdat/calculators/packs.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+import importlib
+from datetime import datetime, timezone
+from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Tuple, Union
+
+from ixdat.db import Saveable
+from ixdat.calculators.indexer import Indexer
+
+
+JsonDict = Dict[str, Any]
+Selector = Union[Iterable[str], Callable[[Any], bool], None]
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py]: interesting type. Is this something that appears often in ixdat?

File: src/ixdat/calculators/packs.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+            raise AttributeError("Measurement has no 'calculators' attribute")
+
+        # Do not include Indexes
+        calculators = [c for c in calculators if not isinstance(c, Indexer)]
+
+        # Apply include/exclude filters
+        if include is not None:
+            calculators = [c for c in calculators if _matches_selector(c, include)]
+        if exclude is not None:
+            calculators = [c for c in calculators if not _matches_selector(c, exclude)]
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py]: I can't figure out why Indexer gets excluded even when I remove the `exclude` argument in the demo script. Can you check if that's also the case for you?

File: src/ixdat/calculators/packs.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+            cal_ver = getattr(cal, "version", None)
+            if cal_ver:
+                item["calculator_version"] = str(cal_ver)
+
+            items.append(item)
+
+        self._items = items
+        return items
+
+    def to_json(self) -> JsonDict:
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py:261]: Ooh, this looks to me like a *third* serialization method :(

File: src/ixdat/calculators/packs.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+        pack.schema_version = str(data.get("schema_version") or cls.SCHEMA_VERSION)
+        pack.ixdat_version = data.get("ixdat_version")
+        pack.created_utc = data.get("created_utc") or _now_utc_iso()
+        pack._metadata = data.get("metadata", {})
+        pack._items = list(data.get("calculators") or [])
+        # Keep canonical db json string if caller wants to save to DB
+        pack.json = json.dumps(data, indent=2)
+        return pack
+
+    def to_file(self, path: str) -> None:
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py]: Probably `export` to be consistent with other Saveable classes that have a way to write themselves to file that isn't through an ixdat backend.

File: src/ixdat/calculators/packs.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+        Read a pack from a file (`path`)
+
+        Returns:
+            CalculatorPack
+        """
+        with open(path, "r", encoding="utf-8") as f:
+            data = json.load(f)
+        return cls.from_json(data)
+
+    def calculators(self) -> List[Any]:
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py:325]: It can be a bit confusing that this is a method whereas `Measurement.calculators` is a dictionary. Not sure which is best practice. Would you suggest changing `Measurement.calculators` to a method to match `CalculatorPack.calculators`?

File: src/ixdat/calculators/packs.py [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+
+            # Always use CalculatorPack’s deserializer
+            cal = self._deserialize_calculator(CalCls, payload)
+            calculators.append(cal)
+
+        self._calculators = calculators
+
+        return list(self._calculators)
+
+    def validate(self, *, strict: bool = True) -> Tuple[bool, List[str]]:
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py:359]: Should all `Saveable` classes have a `validate` method?

File: src/ixdat/calculators/packs.py [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+    ) -> List[Any]:
+        """
+        Attach all calculators in the pack to a Measurement.
+
+        Parameters:
+            measurement: Target Measurement (must provide `.add_calculator(calculator)`).
+            on_conflict:
+                "replace" (default): replace existing calculator of same class+name
+                "skip": keep existing one, skip new
+                "duplicate": add new one and allow duplicates (caller handles names)
```
Comment (ScottSoren) [inline src/ixdat/calculators/packs.py:422]: "duplicate" effectively will require that the calculators can be added. `Measurement.calculators` uses [https://github.com/ixdat/ixdat/blob/40b7ef213d2e013473a500c709d2714de8c6a915/src/ixdat/measurement_base.py#L525](Measurement.consolidate_calculators), which combines calculators of the same type with their `__add__` method. This can make complete sense when e.g. you have an ECCalibration with the electrode area and another ECCalibraiton with the RE potential, or MSCalibrations with sensitivity factors for different molecules. But yeah I guess then they shouldn't have the same names...
So this isn't a request for a change, just something I wanted you to know.

**Review by ScottSoren — CHANGES_REQUESTED** [review]
Nice work!
- Why JSON instead of YAML? Yaml has the advantage that you can make comments in the files. This could be nice if I'm sending someone a text file about the nice calibrations that I've made, which they can load and apply as a CalculatorPack, but also read and see how I got the data. But there could be advantages of JSON over YAML that I don't know.
- The "Indexer" calculator gets excluded even if I don't ask for it to be, and I can't figure out why. I can't think of a case where you would need to save it in a calculator pack, but just bothers me that something doesn't behave
- Other than that, I've left some other comments in the code, with the thorniest I guess questioning the need for multiple serialization/deserialization methods.

**Comment by FrederikLizakJohansen** [comment]
>     * Why JSON instead of YAML? Yaml has the advantage that you can make comments in the files. This could be nice if I'm sending someone a text file about the nice calibrations that I've made, which they can load and apply as a CalculatorPack, but also read and see how I got the data. But there could be advantages of JSON over YAML that I don't know.

Good point! Although my original intent to use JSON comes initially from the fact that (to my knowledge) `pyyaml` isn’t in our dependencies. Therefore JSON is the only format we can rely on everywhere (unless we make some changes to our dependencies ofc.) -- and even if we used YAML, comments wouldn’t survive a load/save cycle anyway. For that reason I think it’s safer to stick with JSON and encourage people to use the notes/metadata fields for any comments they want to keep. I am open to discuss this though.

File: development_scripts/demo_calculator_pack.py (hunk 1 above)
Comment (FrederikLizakJohansen) [reply development_scripts/demo_calculator_pack.py]: Good catch! 
Include=["Indexer"] was acting as a filter, which left only the Indexer. I’ve changed the logic so include is additive (union) instead of restrictive. That means you keep everything else and can explicitly add Indexer back in (or anything else). Default behavior still skips Indexer when no exclude/include is given.

The "ixdat.calculators.indexer.Indexer" was just me testing out the behaviour of the "_matches_selector" helper function. It should work with either. It is now removed from the demo file.

Changes here: c8f15dfbcbde8473b3ca9c6bb4be2edefe1c4363

File: src/ixdat/calculators/packs.py (hunk 6 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: Resolved here: c8f15dfbcbde8473b3ca9c6bb4be2edefe1c4363

File: src/ixdat/calculators/packs.py (hunk 2 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: Good point.
I suggest we keep both: `save()` persists to the configured ixdat backend (IDs, table namespace, discoverable via read(id=...)), while `export()` writes a portable JSON file that can be shared.

Changes here: 3dfc03a868085d8db1cc70c60eb102e4688b6c9d

File: src/ixdat/calculators/packs.py (hunk 3 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: Total valid point. I must concede, I did not think to look deeper into the Saveable class before making these helper functions. It makes much more sense (and makes everything more simple) to reuse the already defined functions and rely on them being correct rather than to add extra redundancy.

I simplified the pack to use the calculators’ own `as_dict()`/`from_dict()` (inherited from Saveable) and removed the extra serializer/deserializer. The pack now just stores class_path + that payload.

For now, let's keep the `as_dict()` serialization for now. We can tackle #75 when I understand ixdat a bit better :-)

Changes here: [7c1b43d](https://github.com/ixdat/ixdat/pull/196/commits/7c1b43db167abbb38dcfc436fe9cbef271d30ec9)

File: src/ixdat/calculators/packs.py (hunk 4 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: See https://github.com/ixdat/ixdat/pull/196#discussion_r2352097803

File: src/ixdat/calculators/packs.py (hunk 5 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: I changed `Selector` → `_Selector` (made it private to `packs.py`) and expanded the `from_measurement` docstring, as well as added a comment above the type definition. Changes here: [8d98714](https://github.com/ixdat/ixdat/pull/196/commits/8d987148d8aed8af61b9c337c918ed4c49d2ab43)

The idea is just to give users flexible ways to filter calculators:
- `None` --> keep all
- list of names --> e.g. `["ECCalibration", "Indexer"]`
- predicate function --> e.g. `lambda c: isinstance(c, ECCalibration) and c.RE_vs_RHE > 0.1`

File: src/ixdat/calculators/packs.py (hunk 8 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: See https://github.com/ixdat/ixdat/pull/196#discussion_r2351614409

File: src/ixdat/calculators/packs.py (hunk 7 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py:269]: I decided to keep `to_json()`/ `from_json() `because they give us a single helper that converts a `CalculatorPack` into a plain JSON-safe dict and back again.

The point is not to introduce a second “serialization layer”, but to have one well-defined transformation between the Python object and the JSON structure we write to file. `export()` and `read()` just wrap these helpers for file I/O.

Changes here: [8ea36dd](https://github.com/ixdat/ixdat/pull/196/commits/8ea36dd550c0997b2ef299ae8a3f2df6503aef26)

**Comment by FrederikLizakJohansen** [comment]
Upon further thinking, I think we can accommodate YAML files, by making pyyaml a soft optional dependency. I have made the changes in f388c00, and switched the default format in the demo to YAML.

File: src/ixdat/calculators/packs.py (hunk 9 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py:368]: Fair. I have changed `calculators()` to a `@property` so `pack.calculators` reads the same as `measurement.calculators`. Changes in f388c00.

File: src/ixdat/calculators/packs.py (hunk 10 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py:402]: I do not think all `Savable` classes need it (at least for now). `CalculatorPack` stores SHA256 hashes per calculator entry, which I believe is unique to its file-exchange use case. I've opened it as an issue in #202.

File: src/ixdat/calculators/packs.py (hunk 11 above)
Comment (FrederikLizakJohansen) [reply src/ixdat/calculators/packs.py]: Thanks, that's a useful clarification. Worth noting in the docstring that "duplicate" bypasses `consolidate_calculators`. So if you want the `__add__`-style merging you're better off calling `measurement.add_calculator()` directly. I've added a note in the docstring in 0508dc8.

**Comment by FrederikLizakJohansen** [comment]
With #209 merged, ixdat can save a measurement and its calculators together in one SQLite file, then load them again. That meets our current need to keep calibrations together, so I’m closing this PR.

I’ll leave #184 open for the remaining idea of transferring a set of calculators to another measurement.
